### PR TITLE
Better avatar animation when attached to a moving entity

### PR DESCRIPTION
**Pull request by hyperlogic**
Added moving-platform.js test script.  When avatar is attached to the moving platform, character motion and animations should properly be relative to the platform.  i.e. If you are standing still on the platform, but moving in the world, your avatar should properly stay in the idle state.

NOTE: This only accurate for non-rotating entities, rotating platform support will be added later.

The SpatiallyNestable class has been extended to support velocity and angularVelocity.

### CODE REVIEW

**Comment by hyperlogic** [comment]
https://app.asana.com/0/45663370065758/83169551830905

**Comment by hifi-gustavo** [comment]
Build console output for commit: 072172b1a204bc97d62ca13466b734ced5b7f483 available at: https://gist.github.com/c09fff7fe87538296693<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7048.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7048.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9a0e206843801d99fda7e0943cab808fd5e7a360 available at: https://gist.github.com/c09fff7fe87538296693<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7048.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7048.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6a82594a4f50463fe9254dc6060fbb9aa4cf8562 available at: https://gist.github.com/c09fff7fe87538296693<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7048.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7048.dmg

**Comment by hyperlogic** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6a82594a4f50463fe9254dc6060fbb9aa4cf8562 available at: https://gist.github.com/c09fff7fe87538296693<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7048.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7048.dmg

File: interface/src/avatar/MyAvatar.cpp [hunk 1]
```diff
@@ -1078,6 +1078,9 @@ void MyAvatar::rebuildCollisionShape() {
 
 void MyAvatar::prepareForPhysicsSimulation() {
     relayDriveKeysToCharacterController();
+
+    bool success;
+    _characterController.setParentVelocity(getParentVelocity(success));
```
Comment (ZappoMan) [inline interface/src/avatar/MyAvatar.cpp]: what happens if getParentVelocity() fails? what value does it return? do you really want to blindly set that? If so, and it's safe, maybe you can change `success` to `safelyIgnore`

**Comment by OmegaHeron** [comment]
Tested against current production build with Windows 7.  Production build, run script and AV "walks" off box getting left behind.  This PR Build AV rides with box - if you fly away AV lets go and returns to expected animation - rejoin the moving box and AV returns to idle stand pose riding along.  Kinda fun/cool actually.

File: libraries/shared/src/SpatiallyNestable.cpp [hunk 2]
```diff
@@ -391,6 +391,77 @@ void SpatiallyNestable::setOrientation(const glm::quat& orientation) {
...
+    _velocityLock.withWriteLock([&] {
+        // TODO: take parent angularVelocity into account.
+        _velocity = glm::inverse(parentTransform.getRotation()) * velocity - parentVelocity;
+    });
+}
+
+void SpatiallyNestable::setVelocity(const glm::vec3& velocity) {
+    bool success;
+    setVelocity(velocity, success);
+}
```
Comment (sethalves) [inline libraries/shared/src/SpatiallyNestable.cpp]: do qDebug here if success is false

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2b10fea00620341c5bb39731e8720309b3072995 available at: https://gist.github.com/c09fff7fe87538296693<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7048.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7048.dmg

**Pull request merged**